irohad/multi_sig_transactions/impl/mst_processor_impl.cpp

```diff
@@ -3,32 +3,89 @@
  * SPDX-License-Identifier: Apache-2.0
  */
 
+#include "multi_sig_transactions/mst_processor_impl.hpp"
+
 #include <utility>
 
+#include <rxcpp/operators/rx-flat_map.hpp>
+#include <rxcpp/operators/rx-map.hpp>
 #include "logger/logger.hpp"
-#include "multi_sig_transactions/mst_processor_impl.hpp"
 
 using shared_model::interface::types::PublicKeyHexStringView;
 
 namespace iroha {
 
   FairMstProcessor::FairMstProcessor(
-      std::shared_ptr<iroha::network::MstTransport> transport,
+      std::shared_ptr<network::MstTransport> transport,
       std::shared_ptr<MstStorage> storage,
       std::shared_ptr<PropagationStrategy> strategy,
       std::shared_ptr<MstTimeProvider> time_provider,
       logger::LoggerPtr log)
       : MstProcessor(log),  // use the same logger in base class
+        log_(std::move(log)),
         transport_(std::move(transport)),
         storage_(std::move(storage)),
         strategy_(std::move(strategy)),
         time_provider_(std::move(time_provider)),
-        propagation_subscriber_(strategy_->emitter().subscribe(
-            [this](auto data) { this->onPropagate(data); })),
-        log_(std::move(log)) {}
+        propagation_subscriber_(
+            strategy_->emitter()
+                .flat_map([](auto data) {
+                  return rxcpp::observable<>::iterate(data).map(
+                      [size = data.size()](auto dst_peer) {
+                        return std::make_tuple(std::move(dst_peer), size);
+                      });
+                })
+                .flat_map(
+                    [log = std::weak_ptr<logger::Logger>(log_),
+                     transport =
+                         std::weak_ptr<network::MstTransport>(transport_),
+                     storage = std::weak_ptr<MstStorage>(storage_),
+                     time_provider = std::weak_ptr<MstTimeProvider>(
+                         time_provider_)](auto tpl)
+                        -> rxcpp::observable<std::tuple<
+                            bool,
+                            std::shared_ptr<shared_model::interface::Peer>,
+                            MstState>> {
+                      auto &[dst_peer, size] = tpl;
+
+                      auto l = log.lock();
+                      auto t = transport.lock();
+                      auto s = storage.lock();
+                      auto tp = time_provider.lock();
+
+                      if (l and t and s and tp) {
+                        auto current_time = tp->getCurrentTime();
+                        auto diff = s->getDiffState(
+                            PublicKeyHexStringView{dst_peer->pubkey()},
+                            current_time);
+                        if (not diff.isEmpty()) {
+                          l->info("Propagate new data[{}]", size);
+                          return t->sendState(*dst_peer, diff)
+                              .map([dst_peer = std::move(dst_peer),
```

Comment:
I know that it will be only called once for now, but generally `map` can call its functor multiple times, and we do not expect it here
```suggestion
                              .take(1).map([dst_peer = std::move(dst_peer),
```